### PR TITLE
[CI] Fix BnB tests

### PR DESCRIPTION
**Pull request by DN6**
# What does this PR do?
Fix a few failures in the BnB test suite.  

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/diffusers/models/transformers/transformer_wan_animate.py [hunk 1]
```diff
@@ -445,10 +445,14 @@ def __call__(
         # B --> batch_size, T --> reduced inference segment len, N --> face_encoder_num_heads + 1, C --> attn.dim
         B, T, N, C = encoder_hidden_states.shape
 
+        # Flatten T and N so the K/V projections see a 3D tensor; BnB int8 matmul only
+        # accepts 2D/3D inputs and would otherwise fail on this 4D activation.
+        encoder_hidden_states = encoder_hidden_states.flatten(1, 2)  # [B, T, N, C] --> [B, T * N, C]
```
Comment (DN6) [inline src/diffusers/models/transformers/transformer_wan_animate.py:450]: @dg845 Small shape change here to make WanAnimate compatible with BnB 8bit kernels.

**Review by dg845 — APPROVED** [review]
Thanks!

File: src/diffusers/models/transformers/transformer_wan_animate.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/models/transformers/transformer_wan_animate.py:450]: Can we confirm it doesn't affect the results on the non-quantization case?

File: tests/models/testing_utils/common.py [hunk 2]
```diff
@@ -205,6 +205,11 @@ def pretrained_model_kwargs(self) -> Dict[str, Any]:
         """Additional kwargs to pass to from_pretrained (e.g., subfolder, variant)."""
         return {}
 
+    @property
+    def torch_dtype(self) -> torch.dtype:
```
Comment (sayakpaul) [inline tests/models/testing_utils/common.py:209]: Where is this leveraged further?

File: tests/models/transformers/test_models_transformer_flux.py [hunk 3]
```diff
@@ -320,6 +320,51 @@ def pretrained_model_name_or_path(self):
 class TestFluxTransformerBitsAndBytes(FluxTransformerTesterConfig, BitsAndBytesTesterMixin):
     """BitsAndBytes quantization tests for Flux Transformer."""
 
+    @property
+    def torch_dtype(self):
+        return torch.float16
+
+    def get_dummy_inputs(self, batch_size: int = 1) -> dict[str, torch.Tensor]:
```
Comment (sayakpaul) [inline tests/models/transformers/test_models_transformer_flux.py]: I think it would be better to have `get_dummy_inputs()` accept a dtype and use it to pass to `randn_tensor()`. Better reusability.

File: tests/models/transformers/test_models_transformer_flux.py [hunk 4]
```diff
@@ -320,6 +320,51 @@ def pretrained_model_name_or_path(self):
 class TestFluxTransformerBitsAndBytes(FluxTransformerTesterConfig, BitsAndBytesTesterMixin):
```
Comment (sayakpaul) [inline tests/models/transformers/test_models_transformer_flux.py:335]: I guess this should be propagated to other tests such as Flux2, Hunuyan, etc.?

**Review by sayakpaul — COMMENTED** [review]
Will fix https://github.com/huggingface/diffusers/pull/13330 accordingly. https://github.com/huggingface/diffusers/pull/13116 is also related.

File: tests/models/testing_utils/common.py (hunk 2 above)
Comment (DN6) [reply tests/models/testing_utils/common.py:209]: It's just meant to explicitly define the default `torch_dtype` for all testing configs.

File: src/diffusers/models/transformers/transformer_wan_animate.py (hunk 1 above)
Comment (DN6) [reply src/diffusers/models/transformers/transformer_wan_animate.py:450]: The flattening only changes the indexing. Since the next op is the qkv projection, it should only affect the last dim. 

This snippet produces zero max diff between the outputs. 

```python
import torch
from diffusers.models.transformers.transformer_wan_animate import (
     WanAnimateFaceBlockCrossAttention,
     WanAnimateFaceBlockAttnProcessor
)

  torch.manual_seed(42)

  dim = 64
  heads = 4
  dim_head = dim // heads
  B, S, T, N, C = 1, 20, 5, 3, dim

  attn = WanAnimateFaceBlockCrossAttention(
      dim=dim, heads=heads, dim_head=dim_head, eps=1e-6,
      cross_attention_dim_head=dim_head,
      processor=WanAnimateFaceBlockAttnProcessor(),
  ).cuda().eval()

  hidden_states = torch.randn(B, S, C, device="cuda")
  encoder_hidden_states = torch.randn(B, T, N, C, device="cuda")

  with torch.no_grad():
      hs = attn.pre_norm_q(hidden_states)
      es = attn.pre_norm_kv(encoder_hidden_states)

      # Old: 4D encoder_hidden_states passed directly to projections
      key_old = attn.to_k(es)
      value_old = attn.to_v(es)

      # New: flatten then project
      es_flat = es.flatten(1, 2)
      key_new = attn.to_k(es_flat)
      value_new = attn.to_v(es_flat)

      # Reshape both to same target shape
      key_old = key_old.view(B, T, N, heads, -1)
      value_old = value_old.view(B, T, N, heads, -1)
      key_new = key_new.view(B, T, N, heads, -1)
      value_new = value_new.view(B, T, N, heads, -1)

      print("key match:", torch.allclose(key_old, key_new, atol=0))
      print("value match:", torch.allclose(value_old, value_new, atol=0))
      print("key max diff:", (key_old - key_new).abs().max().item())
      print("value max diff:", (value_old - value_new).abs().max().item())
```

File: tests/models/testing_utils/common.py (hunk 2 above)
Comment (DN6) [reply tests/models/testing_utils/common.py:209]: We now use it to set the `dtype` explicity in the tester config `get_dummy_inputs()` so then now any testing class that needs to change the dtype can just set it's own `torch_dtype` property and not redefine the `get_dummy_inputs()` method.

File: tests/models/transformers/test_models_transformer_flux.py (hunk 4 above)
Comment (DN6) [reply tests/models/transformers/test_models_transformer_flux.py:335]: Will do in a follow up

File: tests/models/transformers/test_models_transformer_flux.py (hunk 3 above)
Comment (DN6) [reply tests/models/transformers/test_models_transformer_flux.py]: I actually think we should remove all args from get_dummy_inputs and keep it very simple.  

I'll open a follow up PR to remove the batch_size arg from the tests (it's only used in a single parallelism test at the moment).

Comment (sayakpaul) [reply tests/models/transformers/test_models_transformer_flux.py]: And define the same method only varying the batch size, dtype, etc.? No, I don't think we should do that unless I am missing out on something.

File: src/diffusers/models/transformers/transformer_wan_vace.py [hunk 5]
```diff
@@ -331,7 +331,7 @@ def forward(
                 )
                 if i in self.config.vace_layers:
                     control_hint, scale = control_hidden_states_list.pop()
-                    hidden_states = hidden_states + control_hint * scale
+                    hidden_states = hidden_states + control_hint.to(hidden_states.device) * scale
```
Comment (sayakpaul) [inline src/diffusers/models/transformers/transformer_wan_vace.py:334]: Why is this change needed? 👀

File: tests/models/testing_utils/quantization.py [hunk 6]
```diff
@@ -575,33 +567,28 @@ def test_bnb_original_dtype(self):
...
-        if not hasattr(self.model_class, "_keep_in_fp32_modules"):
-            pytest.skip(f"{self.model_class.__name__} does not have _keep_in_fp32_modules")
+        fp32_modules = getattr(self.model_class, "_keep_in_fp32_modules", None)
+        if not fp32_modules:
+            pytest.skip(f"{self.model_class.__name__} does not declare _keep_in_fp32_modules")
 
         config_kwargs = BitsAndBytesConfigMixin.BNB_CONFIGS["4bit_nf4"]
 
-        original_fp32_modules = getattr(self.model_class, "_keep_in_fp32_modules", None)
-        self.model_class._keep_in_fp32_modules = ["proj_out"]
```
Comment (sayakpaul) [inline tests/models/testing_utils/quantization.py:584]: How is this implemented otherwise? It seems to be implemented for models we don't test bitsandbytes for:

https://github.com/search?q=repo%3Ahuggingface%2Fdiffusers%20_keep_in_fp32_modules&type=code

IMO, we could probably make it a property for easier testing or something similar.

File: tests/models/transformers/test_models_transformer_flux.py (hunk 3 above)
Comment (sayakpaul) [reply tests/models/transformers/test_models_transformer_flux.py]: Oh I guess `dummy_input()` is now deriving the `self.torch_dtype` property. Nice. But how to extend to batch sizes? 👀 Okay to do in follow-up.

**Review by sayakpaul — COMMENTED** [review]
I have this comment: https://github.com/huggingface/diffusers/pull/13481/changes#r3092426590

File: src/diffusers/models/transformers/transformer_wan_vace.py (hunk 5 above)
Comment (DN6) [reply src/diffusers/models/transformers/transformer_wan_vace.py:334]: When running `device_map` tests across multiple devices the vace blocks end up on a different device, so the control_hidden_states tensor needs to be moved to the hidden_states tensor in order for inference to work.

File: tests/models/testing_utils/quantization.py (hunk 6 above)
Comment (DN6) [reply tests/models/testing_utils/quantization.py:584]: It exists in the original BnB tests
https://github.com/huggingface/diffusers/blob/c8c84018e0d8704e44d68ff18b634f1f61a717f6/tests/quantization/bnb/test_4bit.py#L235

Here instead of manually adding a dummy keep in fp32 module with `self.model_class._keep_in_fp32_modules = ["proj_out"]` we just use the one that has been defined for the model class if it exists. 

If we use fp16 for inference it will likely fail because the it's artificially injected fp32 module might not handle  the casting properly. The original test seems to place the inference part under an autocast, but that jsut means the keep in FP32 module runs in FP16.

**Review by sayakpaul — APPROVED** [review]
